test/functorch/test_control_flow.py

```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+        dim = 1
+
+        op, op_pt = (add, torch.cumsum)
+        result = scan_fct(op, input=x, dim=dim, reverse=reverse)
+
+        inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+        # Init scalar
+        init = 1.0
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Init wrong shape (More than one time step)
+        init = torch._ops.ops.aten.slice(x, dim, 0, 2, 1)
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Init wrong shape (Other dim different)
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        init = torch.tile(init, (1, 2, 1))
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Init wrong pytree
+        init = (
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+        )
+        with self.assertRaisesRegex(Exception, ".*"):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        # Correct case
+        op, op_pt = (add, torch.cumsum)
+        x = torch.randn(3, 2, 2, device=device)
+        dim = 1
+
+        result = scan_fct(op, input=x, dim=dim, reverse=reverse)
+        result_exp = _fake_scan(op, x, dim=dim, reverse=reverse)
+
+        if reverse:
+            x = torch.flip(x, [dim])
+
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+        result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        self.assertEqual(result, result_exp)
+        self.assertEqual(result_init, result_exp)
+        if not reverse:
+            result_exp_PT = op_pt(x, dim)
+            self.assertEqual(result, result_exp_PT)
+            self.assertEqual(result_init, result_exp_PT)
+
+    @unittest.skipIf(not SM70OrLater, "triton")
+    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_compile_cnt(self, reverse, device):
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        dim = 1
+
+        torch.compiler.reset()
+        from torch._dynamo.testing import CompileCounter
+
+        # Tests rely on automatic_dynamic = True
+        with torch._dynamo.config.patch(automatic_dynamic_shapes=True):
+            cnt = CompileCounter()
+            x = torch.randn(3, 2, 5, device=device)
+            torch.compile(scan, backend=cnt)(add, x, dim=dim, reverse=reverse)
+            self.assertEqual(cnt.frame_count, 1)
+
+            x = torch.randn(3, 20, 5, device=device)
+            # Recompilation
```

Comment:
Please add an explantion for why there is compilation/no recompilation.

Reply:
I added a description of what I think should happen and why recompilations, or no recompilations are done.